top: Generic neopixel support via machine.bitstream.

jimmo opened the pull request:
Extending on the idea from #5294. Instead of `machine.pixelbitstream`, it's `machine.bitstream` with an `encoding` parameter, which allows this to be extended in the future to other modulations. I also provide an ESP32 & ESP8266 implementation, as well as improved the timing for the STM32 version.

Unlike the approach in #5318 (where the bit-banging is common and ports just provide a high-precision delay), this PR allows for better timing accuracy.

I think this also addresses the same issues from #5117 for ESP32 (overflow of the tick counter, and being able to have arbitrary timing tuples). @maludwig, I would be keen to get your feedback.

The `neopixel` module (previously specific to esp32/esp8266) has been updated to use `machine.bitstream` and moved into `drivers` as it is now shared for all ports.

In the future it would be useful to consider optimised (especially non-blocking) implementations for ports (e.g. RMT on ESP32 and DMA SPI on STM32). However, I think the level to implement that is with a different Python-level API (e.g. AsyncNeoPixel) which can be implemented for each port. (See #5184 for an RMT-based implementation).

Tested (both on real neopixels and checking timing with a logic analyser) on ESP32 SPIRAM, ESP32, ESP32S2, ESP8266, STM32 F091, STM32F405 (PYBV11), STM32F7 (PYBD_SF6) all at various clock speeds.

I'd like to leave samd, imx, nrf, rp2 for a future PR, which would abstract the STM32 implementations (and possibly move into extmod/machine_bitstream.c).

jimmo committed:
extmod: Add machine.bitstream.

This is a generic API for synchronously bit-banging data on a pin.

Initially this adds a singled supported encoding, which supports
controlling WS2812 LEDs.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
stm32: Add implementation of machine.bitstream.

Hand-written version for M0, and cycle-counter version for everything else.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
esp32: Replace esp.neopixel with machine.bitstream.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
esp8266: Replace esp.neopixel with machine.bitstream.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
drivers/neopixel: Add common machine.bitstream-based neopixel module.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
stm32: Add neopixel module to STM32.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo committed:
docs: Add docs for machine.bitstream and neopixel.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
@OutOfTheBots I hope this should address your comments in https://github.com/micropython/micropython/pull/5318#issuecomment-620946387 -- It allows the timing (in nanoseconds) to be specified by the user. A timer-based implementation for STM32 would allow for better accuracy, but this approach seems to be working well on F0 and F4.

dpgeorge commented on drivers/neopixel/__init__.py:
```diff
@@ -1,7 +1,11 @@
-# NeoPixel driver for MicroPython on ESP32
-# MIT license; Copyright (c) 2016 Damien P. George
+# NeoPixel driver for MicroPython
```
Was there a reason to have this as `neopixel/__init__.py`, rather than just `neopixel/neopixel.py`?  Ie a single-file module rather than a whole package?  It would be slightly more efficient to import as a single-file, and also use a little less flash and RAM.

jimmo replied:
I was considering making it so you could use the options to `include()` so that you could enable/disable features (exactly like what we do with aioble). But for now just neopixel.py is fine, we can always change it later if necessary. Updated.

dpgeorge replied:
Great, thanks.  At this stage there is no difference to the user having it as a module or package, so module is simpler.

dpgeorge commented on drivers/neopixel/__init__.py:
```diff
@@ -1,7 +1,11 @@
+# MIT license; Copyright (c) 2016 Damien P. George, 2021 Jim Mussared
 
-from esp import neopixel_write
+from machine import bitstream
```
There will need to be a `from micropython import const` here.

dpgeorge commented on ports/esp8266/modmachine.c:
```diff
@@ -413,6 +414,10 @@ STATIC const mp_rom_map_elem_t machine_module_globals_table[] = {
     { MP_ROM_QSTR(MP_QSTR_time_pulse_us), MP_ROM_PTR(&machine_time_pulse_us_obj) },
 
+    #if MICROPY_PY_MACHINE_BITSTREAM
+    { MP_ROM_QSTR(MP_QSTR_bitstream), MP_ROM_PTR(&machine_bitstream_obj) },
```
place just before `time_pulse_us` for consistency

dpgeorge commented on ports/stm32/mpconfigport.h:
```diff
@@ -191,6 +191,7 @@
 #ifndef MICROPY_PY_MACHINE
 #define MICROPY_PY_MACHINE          (1)
 #define MICROPY_PY_MACHINE_PULSE    (1)
+#define MICROPY_PY_MACHINE_BITSTREAM (1)
```
place just before `MICROPY_PY_MACHINE_PULSE` for consistency

dpgeorge commented on ports/stm32/boards/manifest.py:
```diff
@@ -2,3 +2,4 @@
 freeze("$(MPY_DIR)/drivers/dht", "dht.py")
 freeze("$(MPY_DIR)/drivers/display", ("lcd160cr.py", "lcd160cr_test.py"))
 freeze("$(MPY_DIR)/drivers/onewire", "onewire.py")
+include("$(MPY_DIR)/drivers/neopixel/manifest.py")
```
I'm on the fence with adding this to stm32: it is convenient, but stm32 has not needed it so far and it's easy enough to just copy across `neopixel.py` if needed, and including it by default sets a precedence and makes it harder to remove from the build in the future.

jimmo replied:
Yes, that makes sense. Happy to leave it out on STM32 for now.

Speaking of STM32, the CI failed for the NUCLEO_L073RZ board due to flash size (it was fine earlier in the week but this and fstrings pushes it over the limit). I've disabled machine.bitstream on that board. (FWIW I didn't notice this locally because later version of arm-none-eabi-gcc is a few hundred bytes better at size).

jimmo replied:
Updated the branch with neopixel.py no longer added in ports/stm32/boards/manifest.py

mzdaniel commented:
@jimmo: Thank you so much for moving this PR forward and would like to help bringing it to nrf too.

From some quick visual tests I did in esp32, it is working.


For what I saw in this PR:

* Could we add a comment on source code with expected timing order (maybe a reference to the docs)?

* Consistency in variable names on ports, stm32 is different from esp32 (eg: NS_TICKS_OVERHEAD vs NS_CYCLES_OVERHEAD)

* Also, been thinking in the posibility to add neopixel functionality for the nrf using @micropython.asm_thumb insted of in inline C

jimmo commented:
> From some quick visual tests I did in esp32, it is working.

Thanks @mzdaniel 

> 
> For what I saw in this PR:
> 
> * Could we add a comment on source code with expected timing order (maybe a reference to the docs)?

There's a comment in extmod/machine_bitstream.c that explains the timing order for the supported encoding.

```c
// Timing is a 4-tuple of (high_time_0, low_time_0, high_time_1, low_time_1)
// suitable for driving WS2812.
#define MICROPY_MACHINE_BITSTREAM_TYPE_HIGH_LOW (0)
```

Should I move that comment somewhere else?

> * Consistency in variable names on ports, stm32 is different from esp32 (eg: NS_TICKS_OVERHEAD vs NS_CYCLES_OVERHEAD)

This was deliberate, although I'm happy to change it -- on STM32 we're counting cycles (via CYCCNT), on ESP32 we're counting ticks via the `mp_hal_ticks_cpu` API (although technically on ESP32 ticks=cycles, as you can see from the calculation).

> * Also, been thinking in the posibility to add neopixel functionality for the nrf using @micropython.asm_thumb insted of in inline C

This is definitely possible. However for nRF it should just be a matter of porting the stm32 driver, or better (as I suggested in the PR description) to move the STM32 driver into a common location shared by all ARM ports. But as a starting point just duplicating machine_bitstream.c into the nRF port and changing the BSRR register and mask to the correct nRF equivalent (especially for the nRF52, where you can use the cycle counter version).

mzdaniel commented:
> There's a comment in extmod/machine_bitstream.c that explains the timing order for the supported encoding.
> 
> ```c
> // Timing is a 4-tuple of (high_time_0, low_time_0, high_time_1, low_time_1)
> // suitable for driving WS2812.
> #define MICROPY_MACHINE_BITSTREAM_TYPE_HIGH_LOW (0)
> ```
> 
> Should I move that comment somewhere else?
> 
Seems like that comment or a reference to a common place would be more useful in the port implementation files

> > * Consistency in variable names on ports, stm32 is different from esp32 (eg: NS_TICKS_OVERHEAD vs NS_CYCLES_OVERHEAD)
> 
> This was deliberate, although I'm happy to change it -- on STM32 we're counting cycles (via CYCCNT), on ESP32 we're counting ticks via the `mp_hal_ticks_cpu` API (although technically on ESP32 ticks=cycles, as you can see from the calculation).
>
The thing is that different names for the same practical meaning is more taxing on the reader

> > * Also, been thinking in the posibility to add neopixel functionality for the nrf using @micropython.asm_thumb insted of in inline C
> 
> This is definitely possible. However for nRF it should just be a matter of porting the stm32 driver, or better (as I suggested in the PR description) to move the STM32 driver into a common location shared by all ARM ports. But as a starting point just duplicating machine_bitstream.c into the nRF port and changing the BSRR register and mask to the correct nRF equivalent (especially for the nRF52, where you can use the cycle counter version).

Yes, will be fantastic to have as much code shared as posible across ports. Definitelly will try this (I assumed it wont work based on my trials at #5318)

jimmo committed:
docs/library/neopixel.rst: Note that only included on ESP.

Signed-off-by: Jim Mussared <jim.mussared@gmail.com>

jimmo commented:
> Seems like that comment or a reference to a common place would be more useful in the port implementation files

@mzdaniel  I've updated the port implementation files to include a description of the timing array.

> Yes, will be fantastic to have as much code shared as posible across ports. Definitelly will try this (I assumed it wont work based on my trials at #5318)

At 64MHz the cycle counter approach might not work, but the M0 implementation should (as you've demonstrated with the inline assembly).

> I'm on the fence with adding this to stm32

@dpgeorge I've also added a note to the docs (neopixel.rst) to indicate that this module is only available on ESP32/ESP8266 and how to obtain it for STM32.

dpgeorge commented:
Merged in 870000f35b2d278bd759c54ac59e2f97fcdc7dfd through 53145c4c5f10c3e44ffe174b1e6968792f17ea3a with some minor fixes to intermediate commits, and to the docs.

Fantastic work, thank you!
